Create a guide and overview of flatpak extensions (#419)
flatpak/flatpak-docs: docs/extension.rst

  extension to be loaded. Use this if nothing in your buildsystem
  creates that directory. If the directory is missing, there will be
  an error while running the application.

This is a typical example of loading an existing extension from the
Platform or SDK in the application.

.. code-block:: yaml

  id: org.flatpak.cool-app
  runtime: org.kde.Platform
  runtime-version: 5.15-23.08
  sdk: org.kde.Sdk
  command: foo
  add-extensions:
    org.freedesktop.Platform.ffmpeg-full:
      version: '23.08'
      directory: lib/ffmpeg
      add-ld-path: .
  cleanup-commands:
    - mkdir -p ${FLATPAK_DEST}/lib/ffmpeg

Extension manifest
------------------

Once the extension point is defined, an extension like 
``org.flatpak.app.plugin.foo`` can be created.

This is a typical example of such an extension manifest. The explanation
is discussed below.

.. code-block:: yaml

  id: org.flatpak.app.plugin.foo
  branch: '1.0'
  runtime: org.flatpak.app
  runtime-version: 'stable'
  sdk: org.gnome.Sdk//45
  build-extension: true
  separate-locales: false
  appstream-compose: false
  build-options:
    prefix: /app/extension_directory/foo

>     env:
>       PATH: /app/extensions/foo/bin:/app/bin:/usr/bin

Reviewer: ```suggestion
    prepend-path: /app/extensions/foo/bin
```

Reviewer: it's only necessary if you install some binary part of the build process.

similarly `prepend-pkg-config-path` for .pc files from dependencies.

Author: Done, thanks

  modules:
    - name: foo
      buildsystem: simple
      build-commands:
        - <build commands>
        - install -Dm644 --target-directory=${FLATPAK_DEST}/share/metainfo org.flatpak.app.plugin.foo.metainfo.xml

>         - appstream-compose --basename=org.flatpak.app.plugin.foo --prefix=${FLATPAK_DEST} --origin=flatpak org.flatpak.app.plugin.foo

Author: Done, thanks

manifest.

Extension point
---------------

Now to define an extension point the ``add-extensions`` property can be 
used in the manifest.

This is a typical example of an application extension point.

.. code-block:: yaml

  id: org.flatpak.app
  runtime: org.gnome.Platform
  runtime-version: '45'
  sdk: org.gnome.Sdk
  command: foo
  add-extensions:
    org.flatpak.app.plugin:
      version: '1.0'
      directory: extension_directory
      add-ld-path: lib
      merge-dirs: plug-ins;help
      subdirectories: true
      no-autodownload: true
      autodelete: true
  cleanup-commands:
    - mkdir -p ${FLATPAK_DEST}/extension_directory

The details of these properties are documented in the "Extension NAME" 
section of :doc:`flatpak-command-reference`. Some of it is discussed 
further below.

- ``org.flatpak.app.plugin`` is the name of the extension point. 
  Extensions using this extension point must start their ID with the same 
  prefix, for example ``org.flatpak.app.plugin.foo``.

- ``version`` is the branch to use when looking for the extension. This 
  is useful to have parallel installations of the same extension in case 

>   of an API/ABI break. This should match with the ``branch`` used in the 
>   extension manifest.

Reviewer 2: ```suggestion
  of an API/ABI break. This will be the ``branch`` used in the 
  extension manifest.
```

Author: Done, thanks

point or use an existing one. The decision depends on where the extension
should be made available.

If for example it is a GTK or QT theme, it's best to define or use an
extension point in the runtime, so that it can be made available for 
every application utilising said runtime.

If it is a specific tool eg. MPV plugins, which are only used by MPV, 
it's best to define or use an extension point in the application.

To see what are the available extension points in runtime or 
applications, please see the corresponding flatpak or build
manifest.

Extension point
---------------

Now to define an extension point the ``add-extensions`` property can be 
used in the manifest.

This is a typical example of an application extension point.

.. code-block:: yaml

  id: org.flatpak.app
  runtime: org.gnome.Platform
  runtime-version: '45'
  sdk: org.gnome.Sdk
  command: foo
  add-extensions:
    org.flatpak.app.plugin:
      version: '1.0'
      directory: extension_directory
      add-ld-path: lib
      merge-dirs: plug-ins;help
      subdirectories: true
      no-autodownload: true
      autodelete: true
  cleanup-commands:

>     - mkdir -p ${FLATPAK_DEST}/extension_directory

Author: cleanup-commands can be anything that runs during the cleanup phase. Runtimes run various integration commands here.

This is the only way afaik to ensure that something runs after the entirety of the build step in the manifest is done.

>     - mkdir -p ${FLATPAK_DEST}/extension_directory

Reviewer: `cleanup-commands`. It's illogical.

They go in a `post-install`.

>     - mkdir -p ${FLATPAK_DEST}/extension_directory

Reviewer: And just to clarify this can be done a *anytime*. It just need to exist in the final flatpak.

>     - mkdir -p ${FLATPAK_DEST}/extension_directory

Author: post-install means some module/command can rm the empty directory from the flatpak if it is not the last module, cleanup-command prevents that

>     - mkdir -p ${FLATPAK_DEST}/extension_directory

Reviewer: > post-install means some module/command can rm the empty directory from the flatpak if it is not the last module/command

No. Only the top-level cleanup or the module that created it can do that.
And like anything else. They can cleanup the binary too.

>     - mkdir -p ${FLATPAK_DEST}/extension_directory

Author: > No. Only the top-level cleanup or the module that created it can do that.

I don't think that's true? Inside build-commands, the modules have access to sandbox filesystem and anything previously built/installed.

<details>
<summary> example </summary>

```yaml
id: org.flatpak.Hello
runtime: org.freedesktop.Platform
runtime-version: '23.08'
sdk: org.freedesktop.Sdk
command: hello
modules:

  - name: hello
    buildsystem: simple
    build-commands:
      - install -Dm755 hello.sh /app/bin/hello
    sources:
      - type: script
        dest-filename: hello.sh
        commands:
          - echo "Hello world"
    post-install:
      - mkdir -p /app/plugins

  - name: second
    buildsystem: simple
    build-commands:
      - rmdir -v /app/plugins
```

>     - mkdir -p ${FLATPAK_DEST}/extension_directory

Reviewer: Point is it doesn't belong to `cleanup-commands`. It's not a cleanup. It's a setup. We are writing a doc and the doc should be correct. I'm pretty sure if you the person that put it in there did it because it was done elsewhere.

>     - mkdir -p ${FLATPAK_DEST}/extension_directory

Reviewer: ideally the declaration of the extension point should be enough to flatpak-builder to create it, but it doesn't.

>     - mkdir -p ${FLATPAK_DEST}/extension_directory

Author: > I'm pretty sure if you the person that put it in there did it because it was done elsewhere.

You ignored multiple reasons I provided above about why I prefer it to `post-install`...

>     - mkdir -p ${FLATPAK_DEST}/extension_directory

Reviewer: I didn't see any valid one.

Your hypothetical manifest could just as well remove the binaries. or a depenency. It's irrelevant to the logic that something to "setup" doesn't belong in a cleanup.

>     - mkdir -p ${FLATPAK_DEST}/extension_directory

Author: I also see no valid reason to change it. I added that it can be done in post-install, so I cosidered it done.

>     - mkdir -p ${FLATPAK_DEST}/extension_directory

Reviewer: it about at making the documentation correct. that's the whole reason. Hence changing it is needed.

>     - mkdir -p ${FLATPAK_DEST}/extension_directory

Author: It's already correct and works. Only you think that cleanup-commands is meant only to run rm. This is mentioned nowhere and the usage here consistent with what it can do and previous docs https://docs.flatpak.org/en/latest/manifests.html#cleanup

Flatpak also supports `unmanaged extensions`, allowing loading extensions
installed externally into ``/var/lib/flatpak/extension`` and
``$XDG_DATA_HOME/flatpak/extension``. This can be useful to expose 
administrator policies, extensions etc. to Flatpak applications.
The extension point of unmanaged extensions look no different than
their regular counterparts.

An example of an unmanaged extension can be found in browsers such as
`Chromium <https://github.com/flathub/org.chromium.Chromium/blob/dc7f731e7b62199a00bfa3ea3d123ff6d16936dc/org.chromium.Chromium.yaml>`_ 
or `Firefox <https://hg.mozilla.org/mozilla-central/diff/59e57f57dcb73a286822276d02f16e7b17018de6/taskcluster/docker/firefox-flatpak/runme.sh>`_ 
on Flathub.

.. code-block:: yaml

  org.chromium.Chromium.Extension:
    version: '1'
    directory: chromium/extensions
    merge-dirs: extensions;native-messaging-hosts;policies/managed;policies/recommended
    subdirectories: true
    no-autodownload: true
    autodelete: true

The Firefox snippet translates to:

.. code-block:: yaml

  add-extensions:
    org.mozilla.firefox.systemconfig:
      directory: etc/firefox
      no-autodownload: true

  cleanup-commands:

>     - mkdir -p ${FLATPAK_DEST}/etc/firefox

Reviewer: same here.
